src/backend/gporca/libgpos/src/_api.cpp

```diff
@@ -131,41 +131,41 @@ gpos_init(struct gpos_init_params *params)
 {
 	CWorker::abort_requested_by_system = params->abort_requested;
 
-	if (GPOS_OK != gpos::CMemoryPoolManager::Init())
+	GPOS_TRY
 	{
-		return;
-	}
+		gpos::CMemoryPoolManager::Init();
+		gpos::CWorkerPoolManager::Init();
+		gpos::CMessageRepository::Init();
+		gpos::CCacheFactory::Init();
 
-	if (GPOS_OK != gpos::CWorkerPoolManager::Init())
-	{
-		CMemoryPoolManager::GetMemoryPoolMgr()->Shutdown();
-		return;
-	}
+#ifdef GPOS_FPSIMULATOR
+		gpos::CFSimulator::Init();
+#endif	// GPOS_FPSIMULATOR
 
-	if (GPOS_OK != gpos::CMessageRepository::Init())
-	{
-		CWorkerPoolManager::WorkerPoolManager()->Shutdown();
-		CMemoryPoolManager::GetMemoryPoolMgr()->Shutdown();
-		return;
+#ifdef GPOS_DEBUG_COUNTERS
+		CDebugCounter::Init();
+#endif
 	}
-
-	if (GPOS_OK != gpos::CCacheFactory::Init())
+	GPOS_CATCH_EX(ex)
 	{
-		return;
-	}
+#ifdef GPOS_DEBUG_COUNTERS
+		CDebugCounter::Shutdown();
```

Comment:
The gpos_terminate() function is very similar to this catch block. It is possible to rewrite gpos_terminate() and call it here?

Reply:
> Is it possible to rewrite gpos_terminate() and call it here?

gpos_terminate does nothing on release builds, but here I still release the allocated memory in case of errors.

Reply:
Although, I did it anyway.